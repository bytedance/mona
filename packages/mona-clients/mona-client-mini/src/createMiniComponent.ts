import * as React from 'react';
import { CUSTOM_REF } from '@bytedance/mona-shared/dist/constants';

export default function createMiniComponent(name: string) {
  const Component = React.forwardRef(({ children, ...props }, ref) => {
    const newProps: any = { ...props };
    // 自定组件ref比较特殊,约定__ref透传react的ref
    // https://microapp.bytedance.com/docs/zh-CN/mini-app/develop/framework/custom-component/ref/
    newProps[CUSTOM_REF] =
      typeof ref === 'function'
        ? ref
        : (e: any) => {
            ref && (ref.current = e);
          };
    return React.createElement(name, newProps, children);
  });
  Component.displayName = name;
  return Component;
}
