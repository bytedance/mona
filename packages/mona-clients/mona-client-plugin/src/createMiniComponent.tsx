import * as React from 'react';

export function createMiniComponent(name: string) {
  // forwardRef只是为了避免传ref保存
  const Component = React.forwardRef(({ children }, _ref) => <>{children}</>);
  Component.displayName = name;
  return Component;
}
