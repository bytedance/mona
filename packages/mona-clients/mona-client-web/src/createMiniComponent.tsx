import * as React from 'react';

export function createMiniComponent(name: string) {
  // forwardRef 只为避免传ref时，控制台报错。
  const Component = React.forwardRef(({ children }, _ref) => <>{children}</>);
  Component.displayName = name;
  return Component;
}
