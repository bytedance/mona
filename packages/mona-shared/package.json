{
  "name": "@bytedance/mona-shared",
  "displayName": "mona-shared",
  "version": "0.2.33",
  "description": "mona shared utils",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "browser": "./dist/index.web.js",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "dependencies": {
    "@babel/register": "^7.16.5"
  },
  "devDependencies": {
    "@types/node": "^16.4.5",
    "react": "^17.0.2",
    "ttypescript": "^1.5.12",
    "typescript-json-schema": "^0.43.0",
    "typescript-transform-paths": "^2.2.4"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
