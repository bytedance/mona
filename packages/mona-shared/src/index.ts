// please copy web file to index.web.ts
export { default as readConfig } from './readConfig';
export { default as searchScriptFile } from './searchScriptFile';
export { default as formatAppConfig } from './formatAppConfig';
export { default as formatPath } from './formatPath';
export { isClassComponent } from './reactNode';

export * from './domain';
export * from './search';
export * from './constants';
export * from './promisify';
export * from './_transformWebStyle';
export * from './_transformWebClass';
export * from './user';
export * from './request';
