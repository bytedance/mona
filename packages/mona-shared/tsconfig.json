{
  "compilerOptions": {
    "target": "es2017",
    "moduleResolution": "node",
    "strict": true,
    "module": "commonjs",
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "pretty": true,
    "sourceMap": true,
    "declaration": true,
    "plugins": [
      {
        "transform": "typescript-transform-paths"
      },
      {
        "transform": "typescript-transform-paths",
        "afterDeclarations": true
      }
    ],
    "outDir": "dist",
    "baseUrl": ".",
    "paths": {
      "@/*": [
        "src/*"
      ]
    }
  },
  "include": [
    "src"
  ]
}