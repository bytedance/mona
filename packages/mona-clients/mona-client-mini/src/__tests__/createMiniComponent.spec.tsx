import React from 'react';
import TestRenderer from 'react-test-renderer';

import createMiniComponent from '../createMiniComponent';
import { CUSTOM_REF } from '@bytedance/mona-shared';

describe('createMiniComponent', () => {
  test('react ref', () => {
    const customRef = React.createRef();
    const Test = createMiniComponent('test-test');
    const testIns = TestRenderer.create(<Test ref={customRef} />);
    expect(testIns.toJSON()).toMatchSnapshot();
    const refCb = testIns.root.findByType('test-test' as any).props[CUSTOM_REF];
    expect(typeof refCb).toBe('function');
    refCb('c');
    expect(customRef.current).toBe('c');
  });

  test('func ref', () => {
    let current;
    const customRef = (e: any) => {
      current = e;
    };
    const Test = createMiniComponent('test-test');
    const testIns = TestRenderer.create(<Test ref={customRef} />);
    expect(testIns.toJSON()).toMatchSnapshot();

    const refCb = testIns.root.findByType('test-test' as any).props[CUSTOM_REF];
    expect(typeof refCb).toBe('function');
    refCb('c');
    expect(current).toBe('c');
  });
});
