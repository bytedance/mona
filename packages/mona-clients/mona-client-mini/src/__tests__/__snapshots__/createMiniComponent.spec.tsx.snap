// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMiniComponent func ref 1`] = `
<test-test
  __ref={[Function]}
/>
`;

exports[`createMiniComponent react ref 1`] = `
<test-test
  __ref={[Function]}
/>
`;
